GQ Design

MatheMax opened the pull request:
Gauss-Legendre rule implemented

kkmann commented:
okay - almost there ;)

I added a few detailed comments that you might want to address (see 'files changed').

Also, as you can see, the build process failed - i.e. your updated code does not install properly ;) See 
https://travis-ci.org/kkmann/otsd/builds/462845453?utm_source=github_status&utm_medium=notification for build output from Travis CI.
The reason here is that you did not update the DESCRIPTION file which manages package dependencies (see also my line comment on use of library() in packages). 
You can easily activate TravisCI for your fork as well - gives you the chance to see if your changes result in any errors on a remote build system before creating a pull request (u did not get an error on your machine because u have the gaussquad package installed locally ;)) 

You do not need to change any configs to use TravisCi, just set up a free account (using ur existing Github account) and activate it for your fork. The badge in the readme will still point to the build status of this repo if u do not change it manually, but you can check the build status on your own travis CI page.

MatheMax commented:
Hey, I've just updated the program and as far as I see, they are no included in the pull request.
Now I implemented a flexible integration rule. I.e., the user can implement the knots and the weights by himself. We have to define a default setting, I propose GQ with 5 nodes, what do you think?

TravisCI does not work at my computer, but I am getting an error of the form "key can't be retrieved". I'm trying to find out what this means. Does it work at yours now?

kkmann commented on R/GQDesign.R:
```diff
@@ -0,0 +1,81 @@
+    ),
+    contains = "Design")
+
+GQDesign <- function(n1, c1f, c1e, n2_pivots, c2_pivots, knots, weights) {
```
... this means that the constructor only takes 'order' as integer input, constructs the GQ rule via

    rule <- gaussquad::legendre.quadrature.rules(order)[[10]]

kkmann commented on R/GQDesign.R:
```diff
@@ -0,0 +1,81 @@
+
+setGeneric("get_knots", function(d, ...) standardGeneric("get_knots"))
+setMethod("get_knots", signature("GQDesign"),
+    function(d, ...){
```
you can get the knots directly from the rule by `d@rule$x`, just need to scale them appropriately

MatheMax replied:
But isn't it nice to have a function that does this automatically?

kkmann replied:
what do u mean by automatically?
For me, it is all about consistency when u use gaussquad why not use their exact knots as well? In what way is that non-automatic?

kkmann replied:
oh I see, you did exactly what i meant. I was't suggesting that people fiddle with d@rule$x automatically, but yoour previous implementation did not use a rule object since being more general...

kkmann commented on R/GQDesign.R:
```diff
@@ -0,0 +1,81 @@
+setClass("GQDesign", representation(
```
u do not need to store the order of the integration rule (it's in the number of rows of the rule ;))
don't ever store information twice - leads to inconsistencies! So you really only need to store the integration rule! (and it is of class data.frame!)

kkmann replied:
rule is of type data.frame

u do not need to store 'order' - it's nrow(rule) ;)

kkmann commented on R/GQDesign.R:
```diff
@@ -0,0 +1,83 @@
+
+setMethod("n2", signature("GQDesign", "numeric"),
+    function(d, z1, ...) ifelse(z1 < d@c1f | z1 > d@c1e, 0, 1) *
+        pmax(0, approx(get_knots(d), d@n2_pivots, xout = z1, method = "linear", rule = 2)$y) )
```
stats::approx to avoid namespace error

kkmann commented on R/GQDesign.R:
```diff
@@ -0,0 +1,83 @@
+        pmax(0, approx(get_knots(d), d@n2_pivots, xout = z1, method = "linear", rule = 2)$y) )
+
+setMethod("c2", signature("GQDesign", "numeric"),
+    function(d, z1, ...) approx(get_knots(d), d@c2_pivots, xout = z1, method = "linear", rule = 2)$y *
```
same here

kkmann commented on R/GQDesign.R:
```diff
@@ -0,0 +1,83 @@
+        integrand   <- function(z1) eval(s@conditional_score, design, z1, ...) *
+            predictive_pdf(s@conditional_score@prior, z1, n1(design), ...)
+        h           <- (design@c1e - design@c1f) / 2
+        mid_section <- h * sum(design@rule$w * integrand(get_knots(design)))
```
why are you not using the function in gaussquad (asking this for the third time now ;))?

MatheMax replied:
it's included now, sorry ;-)

kkmann commented on R/GQDesign.R:
```diff
@@ -0,0 +1,83 @@
+    function(s, design, ...) {
+        # use design specific implementation tailored to this particular
+        # implementation (Gauss Quadrature N points here)
+        poef <- predictive_cdf(s@conditional_score@prior, design@c1f, n1(design))
```
interface to predictive_pdf has changed, sc FivePointDesign

kkmann commented:
okay, this taking too long - i take what you have so far and try integrate the rest myself ;)
